### PR TITLE
Extend comparison methods to accept different datetime types. (#129)

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Signed-off-by: Yury-Fridlyand <yuryf@bitquilltech.com>

### Description
Functions affected: `=`, `!=`, `>`, `<`, `<=` `>=`.

Added automatic cast rules:
```
TIME -> DATETIME
DATE -> DATETIME
DATETIME -> TIMESTAMP
```

This allows to compare values of any of datetime types.

<details>
<summary>Reverted</summary>

Signatures added:
```
TIME, DATE               -> BOOLEAN
TIME, DATETIME           -> BOOLEAN
TIME, TIMESTAMP          -> BOOLEAN
DATE, TIME               -> BOOLEAN
DATE, DATETIME           -> BOOLEAN
DATE, TIMESTAMP          -> BOOLEAN
DATETIME, TIME           -> BOOLEAN
DATETIME, DATE           -> BOOLEAN
DATETIME, TIMESTAMP      -> BOOLEAN
TIMESTAMP, TIME          -> BOOLEAN
TIMESTAMP, DATE          -> BOOLEAN
TIMESTAMP, DATETIME      -> BOOLEAN
```
</details>

See team review and discussion in https://github.com/Bit-Quill/opensearch-project-sql/pull/129.

#### Test data
```
mysql> show fields from calcs where field IN ('date0', 'time0', 'time1', 'datetime0');
+-----------+-----------+------+-----+---------+-------+
| Field     | Type      | Null | Key | Default | Extra |
+-----------+-----------+------+-----+---------+-------+
| date0     | date      | YES  |     | NULL    |       |
| time0     | datetime  | YES  |     | NULL    |       |
| time1     | time      | YES  |     | NULL    |       |
| datetime0 | timestamp | YES  |     | NULL    |       |
+-----------+-----------+------+-----+---------+-------+
mysql> select `key`, date0, time0, time1, datetime0 from calcs limit 7;
+-------+------------+---------------------+----------+---------------------+
| key   | date0      | time0               | time1    | datetime0           |
+-------+------------+---------------------+----------+---------------------+
| key00 | 2004-04-15 | 1899-12-30 21:07:32 | 19:36:22 | 2004-07-09 10:17:35 |
| key01 | 1972-07-04 | 1900-01-01 13:48:48 | 02:05:25 | 2004-07-26 12:30:34 |
| key02 | 1975-11-12 | 1900-01-01 18:21:08 | 09:33:31 | 2004-08-02 07:59:23 |
| key03 | 2004-06-04 | 1900-01-01 18:51:48 | 22:50:16 | 2004-07-05 13:14:20 |
| key04 | 2004-06-19 | 1900-01-01 15:01:19 | NULL     | 2004-07-28 23:30:22 |
| key05 | NULL       | 1900-01-01 08:59:39 | 19:57:33 | 2004-07-22 00:30:23 |
| key06 | NULL       | 1900-01-01 07:37:48 | NULL     | 2004-07-28 06:54:50 |
+-------+------------+---------------------+----------+---------------------+
```
##### MySQL
```sql
select `key`, date0, time0, time1, datetime0, date0 < now(), time0 < now(), time1 < now(), datetime0 < now() from calcs limit 6;
select `key`, date0, time0, time1, datetime0, date0 <= now(), time0 <= now(), time1 <= now(), datetime0 <= now() from calcs limit 6;
```
##### OpenSearch
```sql
SELECT CAST(date0 AS date)                  AS `date`,       CAST(date0 AS date)             < now() AS `date < datetime`,
       datetime(CAST(time0 AS STRING))      AS `datetime`,   datetime(CAST(time0 AS STRING)) < now() AS `datetime < datetime`,
       CAST(time1 AS time)                  AS `time`,       CAST(time1 AS time)             < now() AS `time < datetime`,
       CAST(datetime0 AS timestamp)         AS `timestamp`,  CAST(datetime0 AS timestamp)    < now() AS `timestamp < datetime`
FROM calcs LIMIT 6;
   
SELECT CAST(date0 AS date)                  AS `date`,       CAST(date0 AS date)             <= now() AS `date <= datetime`,
       datetime(CAST(time0 AS STRING))      AS `datetime`,   datetime(CAST(time0 AS STRING)) <= now() AS `datetime <= datetime`,
       CAST(time1 AS time)                  AS `time`,       CAST(time1 AS time)             <= now() AS `time <= datetime`,
       CAST(datetime0 AS timestamp)         AS `timestamp`,  CAST(datetime0 AS timestamp)    <= now() AS `timestamp <= datetime`
FROM calcs LIMIT 6;
```

### Issues Resolved
Fix comparison between different datetime data types. [#294](https://github.com/opensearch-project/sql/issues/294)
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by penghuo — APPROVED** [review]

**Review by dai-chen — CHANGES_REQUESTED** [review]
Sorry to point this out late. I just realized our current implicit cast mechanism may handle this more properly. Its basic idea is to wrap cast function on argument which has different type than resolved argument but compatible. In this way, all functions will cast the argument automatically instead of only comparison operator. Here is the design doc: https://github.com/opensearch-project/sql/blob/main/docs/dev/TypeConversion.md

Code changes in 2 places for test:
1. Make `DATE` and `TIME` compatible with `TIMESTAMP` for our test
2. Add missing `cast_to_timestamp(date)` function impl

```
diff --git a/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java b/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java
index ad2e4259..58a60779 100644
--- a/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java
+++ b/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java
@@ -55,10 +55,10 @@ public enum ExprCoreType implements ExprType {
    * Date.
    * Todo. compatible relationship.
    */
-  TIMESTAMP(STRING),
   DATE(STRING),
   TIME(STRING),
   DATETIME(STRING),
+  TIMESTAMP(DATE, TIME),
   INTERVAL(UNDEFINED),

   /**
diff --git a/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperator.java b/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperator.java
index 8f904bfb..637905a5 100644
--- a/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperator.java
+++ b/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperator.java
@@ -182,6 +182,8 @@ public class TypeCastOperator {
             (v) -> new ExprTimestampValue(v.stringValue())), TIMESTAMP, STRING),
         impl(nullMissingHandling(
             (v) -> new ExprTimestampValue(v.timestampValue())), TIMESTAMP, DATETIME),
+        impl(nullMissingHandling(
+            (v) -> new ExprTimestampValue(v.timestampValue())), TIMESTAMP, DATE),
         impl(nullMissingHandling((v) -> v), TIMESTAMP, TIMESTAMP)
     );
   }
```

My test: (The reason it return true is: `ExprDateValue.timestampValue()` seems consider timezone and return `2020-09-16 07:00:00` for date value `2020-09-16`. I don't see other issue other than this. Please let me know if this can work for you. Thanks!

```
curl -XPOST "localhost:9200/_plugins/_sql" -H 'Content-Type: application/json' -d'
{
  "query": "SELECT DATE('\''2020-09-16'\'') > TIMESTAMP('\''2020-09-16 00:00:00'\'') "
}'
{
  "schema": [
    {
      "name": "DATE('2020-09-16') > TIMESTAMP('2020-09-16 00:00:00')",
      "type": "boolean"
    }
  ],
  "datarows": [
    [
      true
    ]
  ],
  "total": 1,
  "size": 1,
  "status": 200
}
```

File: core/src/main/java/org/opensearch/sql/data/model/ExprValue.java [hunk 1]
```diff
@@ -60,6 +60,15 @@ default boolean isNumber() {
     return false;
   }
 
+  /**
+   * Is Number value.
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/data/model/ExprValue.java:64]: The falcon's eye! Right, thank you.

**Comment by Yury-Fridlyand** [comment]
@dai-chen, thank you for detailed feedback!

I'll try to rework it according to you notes.
So far I see those things:
1. `TypeCastOperator` should be extended to support `TIME` -> `TIMESTAMP` cast which requires `FunctionProperties`
2. Same casts should be added for X -> `DATETIME` (optional, because X -> `TIMESTAMP` -> `DATETIME` is possible)
3. `LuceneQuery` should be updated to reflect changes in `TypeCastOperator`
4. `ExprCoreType` should be like this:
```java
  DATE(STRING),
  TIME(STRING),
  DATETIME(STRING, DATE, TIME),
  TIMESTAMP(STRING, DATE, TIME),
  INTERVAL(UNDEFINED),
```

**Comment by Yury-Fridlyand** [comment]
@dai-chen,
I tried to apply your proposal and did a performance measurements. It has a bit worse performance, unfortunately.
Also it is impossible to update `castMap` in `LuceneQuery` - `TIME` -> `X` cast requires `FunctionProperties`, but `castMap` isn't compatible with it.
Please, see info about performance testing below:
First performance measurement point
```diff
diff --git a/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchQueryManager.java b/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearc>
index 9c6fcdb82..0b2629562 100644
--- a/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchQueryManager.java
+++ b/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchQueryManager.java
@@ -30,7 +30,12 @@ public class OpenSearchQueryManager implements QueryManager {

   @Override
   public QueryId submit(AbstractPlan queryPlan) {
-    schedule(nodeClient, () -> queryPlan.execute());
+    schedule(nodeClient, () -> {
+      long timeBefore = System.currentTimeMillis();
+      queryPlan.execute();
+      long elapsed = System.currentTimeMillis() - timeBefore;
+      System.out.println(String.format("\u001B[31m===== %d =====\u001B[0m", elapsed));
+    });

     return queryPlan.getQueryId();
   }
```
Second:
```diff
   diff --git a/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchExecutionEngine.java b/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSe>
index 9a136a3be..98bffb85c 100644
--- a/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchExecutionEngine.java
+++ b/opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchExecutionEngine.java
@@ -41,7 +41,7 @@ public class OpenSearchExecutionEngine implements ExecutionEngine {
         () -> {
           try {
             List<ExprValue> result = new ArrayList<>();
-
+            long timeBefore = System.currentTimeMillis();
             context.getSplit().ifPresent(plan::add);
             plan.open();

@@ -50,6 +50,8 @@ public class OpenSearchExecutionEngine implements ExecutionEngine {
             }

             QueryResponse response = new QueryResponse(physicalPlan.schema(), result);
+            long elapsed = System.currentTimeMillis() - timeBefore;
+            System.out.println(String.format("\u001B[31m===== %d =====\u001B[0m", elapsed));
             listener.onResponse(response);
           } catch (Exception e) {
             listener.onFailure(e);
```
   
Changes I applied:
```diff
diff --git a/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java b/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java
index ad2e42596..fc964d205 100644
--- a/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java
+++ b/core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java
@@ -55,10 +55,16 @@ public enum ExprCoreType implements ExprType {
    * Date.
    * Todo. compatible relationship.
    */
-  TIMESTAMP(STRING),
   DATE(STRING),
   TIME(STRING),
+/*
   DATETIME(STRING),
+  TIMESTAMP(STRING),
+//*/
+//*
+  DATETIME(STRING, DATE, TIME),
+  TIMESTAMP(STRING, DATE, TIME, DATETIME),
+//*/
   INTERVAL(UNDEFINED),

   /**
diff --git a/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperator.java b/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOpe>
index 8f904bfbf..46e178aed 100644
--- a/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperator.java
+++ b/core/src/main/java/org/opensearch/sql/expression/operator/convert/TypeCastOperator.java
@@ -19,7 +19,9 @@ import static org.opensearch.sql.data.type.ExprCoreType.STRING;
 import static org.opensearch.sql.data.type.ExprCoreType.TIME;
 import static org.opensearch.sql.data.type.ExprCoreType.TIMESTAMP;
 import static org.opensearch.sql.expression.function.FunctionDSL.impl;
+import static org.opensearch.sql.expression.function.FunctionDSL.implWithProperties;
 import static org.opensearch.sql.expression.function.FunctionDSL.nullMissingHandling;
+import static org.opensearch.sql.expression.function.FunctionDSL.nullMissingHandlingWithProperties;

 import java.util.Arrays;
  import java.util.stream.Collectors;
@@ -182,6 +184,12 @@ public class TypeCastOperator {
             (v) -> new ExprTimestampValue(v.stringValue())), TIMESTAMP, STRING),
         impl(nullMissingHandling(
             (v) -> new ExprTimestampValue(v.timestampValue())), TIMESTAMP, DATETIME),
+//*
+        impl(nullMissingHandling(
+            (v) -> new ExprTimestampValue(v.timestampValue())), TIMESTAMP, DATE),
+        implWithProperties(nullMissingHandlingWithProperties(
+            (fp, v) -> new ExprTimestampValue(((ExprTimeValue)v).timestampValue(fp))), TIMESTAMP, TIME),
+//*/
         impl(nullMissingHandling((v) -> v), TIMESTAMP, TIMESTAMP)
     );
   }
@@ -192,6 +200,11 @@ public class TypeCastOperator {
             (v) -> new ExprDatetimeValue(v.stringValue())), DATETIME, STRING),
         impl(nullMissingHandling(
             (v) -> new ExprDatetimeValue(v.datetimeValue())), DATETIME, TIMESTAMP),
+//*
+        implWithProperties(nullMissingHandlingWithProperties(
+            (fp, v) -> new ExprDatetimeValue(((ExprTimeValue)v).datetimeValue(fp))), DATETIME, TIME),
+        impl(nullMissingHandling((v) -> v), DATETIME, DATETIME),
+//*/
         impl(nullMissingHandling(
             (v) -> new ExprDatetimeValue(v.datetimeValue())), DATETIME, DATE)
     );
diff --git a/core/src/main/java/org/opensearch/sql/expression/operator/predicate/BinaryPredicateOperator.java b/core/src/main/java/org/opensearch/sql/expression/operator/predicate/>
index 23547e8d4..5baaa357b 100644
--- a/core/src/main/java/org/opensearch/sql/expression/operator/predicate/BinaryPredicateOperator.java
+++ b/core/src/main/java/org/opensearch/sql/expression/operator/predicate/BinaryPredicateOperator.java
@@ -202,6 +202,13 @@ public class BinaryPredicateOperator {
   private static DefaultFunctionResolver compareImpl(
       FunctionName function, SerializableBiFunction<Comparable, Comparable, Boolean> comparator) {
     return FunctionDSL.define(function,
+//*
+            ExprCoreType.coreTypes().stream()
+                .map(type -> impl(nullMissingHandling(
+                    (v1, v2) -> ExprBooleanValue.of(comparator.apply(v1, v2))),
+                        BOOLEAN, type, type))
+//*/
+/*
         Stream.concat(
             ExprCoreType.coreTypes().stream()
                 .map(type -> impl(nullMissingHandling(
@@ -212,6 +219,7 @@ public class BinaryPredicateOperator {
                     (fp, v1, v2) -> ExprBooleanValue.of(comparator.apply(
                             extractDateTime(v1, fp), extractDateTime(v2, fp)))),
                         BOOLEAN, pair.getLeft(), pair.getRight())))
+//*/
         .collect(Collectors.toList()));
   }
```
   
To test I swapped comments by switching `//*` to `/*` and run different queries. Each time I restart the cluster before making a test.
```SQL
SELECT CAST(date0 AS date)                  AS `date`,       CAST(date0 AS date)             < now() AS `date < datetime`,
       datetime(CAST(time0 AS STRING))      AS `datetime`,   datetime(CAST(time0 AS STRING)) < now() AS `datetime < datetime`,
       CAST(time1 AS time)                  AS `time`,       CAST(time1 AS time)             < now() AS `time < datetime`,
       CAST(datetime0 AS timestamp)         AS `timestamp`,  CAST(datetime0 AS timestamp)    < now() AS `timestamp < datetime`
FROM calcs LIMIT 6;
```
(there is only one time measurement point)
```
chen: 435, 17, 14, 13, 14
yury: 532, 16, 15, 14, 14
```

-------
```sql
select count(*), timestamp > '2018-01-21 05:32:25' from flights group by 2;
```
```
yury: 362, 88, 52, 82, 41
      765, 96, 57, 87, 45
chen: 342, 97, 65, 92, 47
      743, 103, 70, 97, 51
```

-----
```sql
select count(*), timestamp > TIME '05:32:25' from flights group by 2;
```
```
yury: 288, 68, 38, 27, 28
      728, 75, 43, 32, 33
chen: 294, 47, 47, 35, 30
      831, 55, 52, 40, 34
```

-----
```sql
select count(*), timestamp > DATE '2018-01-21' from flights group by 2;
```
```
yury: 268, 41, 37, 23, 23
      656, 48, 41, 28, 27
chen: 259, 58, 28, 30, 26
      658, 6
5, 32, 34, 30
```

------
```sql
select count(*), datetime0 > '2004-07-14 08:16:44' from calcs group by 2;
```
```
yury: 175, 15, 9, 10, 7
      576, 21, 14, 14, 11
chen: 152, 12, 9, 9, 8
      541, 19, 15, 13, 12
```

-------
```sql
select count(*), datetime0 > TIME '08:16:44' from calcs group by 2;
```
```
yury: 187, 11, 8, 7, 6
      585, 19, 13, 12, 9
chen: 139, 12, 8, 7, 7
      470, 18, 12, 11, 11
```

------
```sql
select count(*), datetime0 > DATE '2004-07-14' from calcs group by 2;
```
```
yury: 149, 12, 7, 7, 8
      511, 19, 11, 10, 11
chen: 170, 18, 7, 7, 8
      642, 25, 12, 12, 12
```

**Comment by dai-chen** [comment]
@Yury-Fridlyand Thanks for sharing the test result! I'm not sure if the dominant factor is nested lambda evaluation on argument or resolving `cast` function per query. Because for now I think we don't have performance requirement and the existing mechanism can simplify our code significantly, we may be able to add cache mechanism in function repository or other optimization as needed. I'm looking into the code from my side and will update if any finding. Thanks!

**Comment by dai-chen** [comment]
@Yury-Fridlyand I just did some test by JMH which can microbenchmark on fine-grain class level. Here is my branch and test result: https://github.com/dai-chen/sql-1/tree/test-mircro-benchmark. I think the result is acceptable and it would be very close if we ignore time spent on function compile. I did test quick and not sure if I miss anything. We can double check and let me know if any question. Thanks!

```
Approach 1: Existing Implicit Cast

Benchmark                                                               (dateTimeValue)  (dateValue)  Mode  Cnt   Score    Error  Units
TypeCastOperatorBenchmark.testDateEqualsDateTimeCompileAndEvaluate  2022-01-01 00:00:00   2022-01-01  avgt    3   0.006 ±  0.003  ms/op
TypeCastOperatorBenchmark.testDateEqualsDateTimeEvaluate            2022-01-01 00:00:00   2022-01-01  avgt    3  ≈ 10⁻⁵           ms/op

Approach 2: Current PR Approach

Benchmark                                                               (dateTimeValue)  (dateValue)  Mode  Cnt   Score    Error  Units
TypeCastOperatorBenchmark.testDateEqualsDateTimeCompileAndEvaluate  2022-01-01 00:00:00   2022-01-01  avgt    3   0.004 ±  0.002  ms/op
TypeCastOperatorBenchmark.testDateEqualsDateTimeEvaluate            2022-01-01 00:00:00   2022-01-01  avgt    3  ≈ 10⁻⁵           ms/op
```

**Comment by Yury-Fridlyand** [comment]
Great job @dai-chen! Thank you for sharing your test results.
I'll proceed with your proposed approach.

File: docs/user/dql/expressions.rst [hunk 2]
```diff
@@ -146,6 +146,10 @@ Here is an example for different type of comparison operators::
     | True    | True     | False   | True     | False    | False   |
     +---------+----------+---------+----------+----------+---------+
 
+It is possible to compare datetimes. When comparing different datetime types, for example `DATE` and `TIME`, both converted to `DATETIME`.
+The following rule is applied on coversion: a `TIME` applied to today's date; `DATE` is interpreted at midnight.
+
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply docs/user/dql/expressions.rst:155]: Added in e86e7036d

File: core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java [hunk 3]
```diff
@@ -53,12 +53,11 @@ public enum ExprCoreType implements ExprType {
...
   /**
    * Date.
-   * Todo. compatible relationship.
    */
-  TIMESTAMP(STRING),
   DATE(STRING),
   TIME(STRING),
-  DATETIME(STRING),
+  DATETIME(STRING, DATE, TIME),
+  TIMESTAMP(STRING, DATE, TIME, DATETIME),
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/data/type/ExprCoreType.java]: Fixed in 50fbdc07a

**Review by dai-chen — APPROVED** [review]
LGTM. Thanks for the changes!

**Review by penghuo — APPROVED** [review]

**Comment by dai-chen** [comment]
Since we enhanced CAST operator in this PR, can https://github.com/opensearch-project/sql/issues/853 be linked and closed as well?

**Comment by Yury-Fridlyand** [comment]
> Since we enhanced CAST operator in this PR, can #853 be linked and closed as well?

No,
#853 is about unsupported `DATETIME` in `cast` in V2. It is supported in V1, so #853 is a part of legacy deprecation.
```
opensearchsql> select cast('1961-04-12 09:07:00' as DATETIME);
{'reason': 'Invalid SQL query', 'details': '', 'type': 'NullPointerException'}
```

**Comment by dai-chen** [comment]
> > Since we enhanced CAST operator in this PR, can #853 be linked and closed as well?
> 
> No, #853 is about unsupported `DATETIME` in `cast` in V2. It is supported in V1, so #853 is a part of legacy deprecation.
> 
> ```
> opensearchsql> select cast('1961-04-12 09:07:00' as DATETIME);
> {'reason': 'Invalid SQL query', 'details': '', 'type': 'NullPointerException'}
> ```

You mean with your changes in this PR below, `CAST(string as datetime)` still not work in v2?

```
  private static DefaultFunctionResolver castToDatetime() {
    return FunctionDSL.define(BuiltinFunctionName.CAST_TO_DATETIME.getName(),
        impl(nullMissingHandling(
            (v) -> new ExprDatetimeValue(v.stringValue())), DATETIME, STRING),
        impl(nullMissingHandling(
            (v) -> new ExprDatetimeValue(v.datetimeValue())), DATETIME, TIMESTAMP),
        impl(nullMissingHandling(
            (v) -> new ExprDatetimeValue(v.datetimeValue())), DATETIME, DATE),
        implWithProperties(nullMissingHandlingWithProperties(
            (fp, v) -> new ExprDatetimeValue(((ExprTimeValue)v).datetimeValue(fp))),
            DATETIME, TIME),
        impl(nullMissingHandling((v) -> v), DATETIME, DATETIME)
    );
```

**Comment by Yury-Fridlyand** [comment]
> Is it because grammar changes still missing?

Exactly, grammar fix is required

**Comment by dai-chen** [comment]
> > Is it because grammar changes still missing?
> 
> Exactly, grammar fix is required

I see. Thanks!

**Pull request merged**